src/Drupal/Migrate/MigrateExecutable.php

```diff
@@ -213,6 +213,22 @@ public function onPreImport(MigrateImportEvent $event): void
         $this->initProgressBar($migration);
     }
 
+    /**
+     * The array key to use to represent Source ID values.
+     *
+     * Helps speed up array searched by serializing the value for a key.
+     *
+     * @param array $source_id
+     *   The Source ID array.
+     *
+     * @return string
+     *   The array key to represent Source ID values.
+     */
+    protected function getSourceIdKey(array $source_id): string
+    {
+        return implode('-', $source_id);
```

Comment:
I wonder how safe is this

Reply:
I don't know the full extent of what is possible to appear in the `$source_id` array values, but from my investigation it is mostly entity IDs, revision IDs, and simple strings. Of which, most cases involve one or two values.

I am open to alternate ways of serializing the array values into a unique key.

Reply:
Hi.
I run into this issue while doing a diagnostic on what was slowing the import even after it was marked as finished.

And want to put my grain of salts.

the case i can see where the proposed serialisation can cause problem is very specific.
[A_B, C] has the same key than [A,B_C].

you could use hash function but you would still run into the [Pigeon hole principle](https://en.m.wikipedia.org/wiki/Pigeonhole_principle).

the only way to avoid that would be to let the user decide by a cli argument including the option to turn this off and use the old system.